File: src/main/java/com/baulsupp/oksocial/util/OpenSCUtil.java
package com.baulsupp.oksocial.util;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.lang.reflect.Constructor;
import java.nio.charset.StandardCharsets;
import java.security.KeyStore;
import java.security.Provider;
import java.security.Security;
import javax.net.ssl.KeyManager;
import javax.net.ssl.KeyManagerFactory;

public class OpenSCUtil {
  public static KeyManager[] getKeyManagers(char[] password, int slot) throws Exception {
    String config =
        "name=OpenSC\nlibrary=/Library/OpenSC/lib/opensc-pkcs11.so\nslot=" + slot + "\n";

    Provider pkcs11;

    KeyStore keystore;
    if (isJava9()) {
      pkcs11 = Security.getProvider("SunPKCS11");

      pkcs11 =
          (Provider) Provider.class.getMethod("configure", String.class).invoke(pkcs11, "--" + config);

      Security.addProvider(pkcs11);

      keystore = KeyStore.getInstance("PKCS11", pkcs11);
    } else {
      Constructor<?> ctor =
          Class.forName("sun.security.pkcs11.SunPKCS11").getConstructor(InputStream.class);
      pkcs11 =
          (Provider) ctor.newInstance(
              new ByteArrayInputStream(config.getBytes(StandardCharsets.UTF_8)));

      Security.addProvider(pkcs11);

      keystore = KeyStore.getInstance("PKCS11", pkcs11);
    }

    keystore.load(null, password);

    KeyManagerFactory kmf = KeyManagerFactory.getInstance("NewSunX509");
    kmf.init(keystore, null);

    return kmf.getKeyManagers();
  }

  private static boolean isJava9() {
    return "9".equals(System.getProperty("java.specification.version"));
  }
}

File: src/main/kotlin/com/baulsupp/okurl/security/OpenSCUtil.kt
package com.baulsupp.okurl.security

import java.security.KeyStore
import java.security.Security
import java.util.Arrays
import javax.net.ssl.KeyManagerFactory
import javax.net.ssl.KeyStoreBuilderParameters
import javax.net.ssl.X509ExtendedKeyManager
import javax.security.auth.callback.CallbackHandler

object OpenSCUtil {
  fun getKeyManager(
    callbackHandler: CallbackHandler,
    slot: Int = 0
  ): X509ExtendedKeyManager {
    val config = "--name=OpenSC\nlibrary=/Library/OpenSC/lib/opensc-pkcs11.so\nslot=$slot\n"

    // May fail with ProviderException with root cause like
    // sun.security.pkcs11.wrapper.PKCS11Exception: CKR_SLOT_ID_INVALID
    val pkcs11 = Security.getProvider("SunPKCS11").configure(config)
    Security.addProvider(pkcs11)

    val builderList: List<KeyStore.Builder> = Arrays.asList(
      KeyStore.Builder.newInstance("PKCS11", null, KeyStore.CallbackHandlerProtection(callbackHandler))
    )

    val keyManagerFactory = KeyManagerFactory.getInstance("NewSunX509")
    keyManagerFactory.init(KeyStoreBuilderParameters(builderList))
    return keyManagerFactory.keyManagers[0] as X509ExtendedKeyManager
  }
}
